File: src/main/java/com/sica/backend/repository/SubjectRepository.java
package com.sica.backend.repository;

import com.sica.backend.entity.Subject;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface SubjectRepository extends JpaRepository<Subject, String> {
    Optional<Subject> findByCode(String code);
    List<Subject> findByProfessorId(String professorId);
    List<Subject> findByActiveTrueAndProfessorId(String professorId);

    @Query("SELECT s FROM Subject s JOIN SubjectEnrollment se ON se.subject.id = s.id WHERE se.student.personId = :studentId AND s.active = true")
    List<Subject> findByStudentId(String studentId);
}

File: src/main/kotlin/com/sica/backend/repository/SubjectRepository.kt
package com.sica.backend.repository

import com.sica.backend.entity.Subject
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface SubjectRepository : JpaRepository<Subject, String> {
    fun findByCode(code: String): Optional<Subject>
    fun findByProfessorId(professorId: String): List<Subject>
    fun findByActiveTrueAndProfessorId(professorId: String): List<Subject>
    
    @Query("SELECT s FROM Subject s JOIN SubjectEnrollment se ON se.subject.id = s.id WHERE se.student.personId = :studentId AND s.active = true")
    fun findByStudentId(studentId: String): List<Subject>
}
